ops/relation.py

```diff
@@ -104,49 +110,78 @@ def _on_consumer_joined(self, event):
             logger.debug("Providing for joined consumer : %s", data)
             event.relation.data[self.model.app]['provider_data'] = json.dumps(data)
 
-    def _on_upgrade(self, event):
+    def _on_upgrade(self, event: UpgradeCharmEvent) -> None:
         """Handle a provider upgrade event.
 
         Args:
             event: event object
         """
         self._notify_consumers()
 
-    def _notify_consumers(self):
+    def _notify_consumers(self) -> None:
         """Resend provider data to consumers."""
         data = self._provider_data()
         if self.model.unit.is_leader():
             logger.debug("Notifying Consumer : %s", data)
             for rel in self.framework.model.relations[self.name]:
                 rel.data[self.model.app]['provider_data'] = json.dumps(data)
 
-    def ready(self):
+    def ready(self) -> None:
         """Set provider state to ready."""
         if not self.is_ready:
             logger.debug("Provider is ready")
             self._stored.ready = True
             self._notify_consumers()
 
-    def unready(self):
+    def unready(self) -> None:
         """Set provider state to unready."""
         logger.debug("Provider is not ready")
         self._stored.ready = False
         self._notify_consumers()
 
-    def _provider_data(self):
+    def _provider_data(self) -> typing.Dict:
         """Construct relation data packet for consumer."""
         data = dict()
         data['provides'] = self.provides.copy()
         data['ready'] = self._stored.ready
         return data
 
     @property
-    def is_ready(self):
+    def is_ready(self) -> bool:
         """Query state of provider."""
         return self._stored.ready
 
 
-class ProviderAvailable(EventBase):
+class ProviderEvent(EventBase):
+    """A base class for Provider events.
+
+    Args:
+        handle: a :class:`ops.framework.Handle` to reference the event
+        data: an optional dict containing data
+    """
+    def __init__(self, handle: Handle, relation: Relation,
+                 data: typing.Optional[typing.Dict] = None):
+        super().__init__(handle)
+        self.relation = relation
+        self.data = data
```

Comment:
On second thought would it be more consistent if we nested data under relation `event.relation.data` ? I do appreciate data was not nested previously and it may not be directly related to this PR but if you agree then might as well ...